Userland/Utilities/shred.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2024, Romain Chardiny <romain.chardiny@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <LibCore/System.h>
+#include <LibFileSystem/FileSystem.h>
+#include <stdio.h>
+#include <unistd.h>
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    TRY(Core::System::pledge("stdio unix rpath wpath cpath"));
+
+    Vector<StringView> paths;
+    bool remove_file = false;
+    bool verbose = false;
+    u32 iterations = 3;
+    Optional<StringView> random_source;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_option(remove_file, "Deallocate and remove file after overwriting", nullptr, 'u');
+    args_parser.add_option(verbose, "Show progress", "verbose", 'v');
+    args_parser.add_option(iterations, "Overwrite N times instead of the default (3)", "iterations", 'n', "N");
+    args_parser.add_option(random_source, "Get random bytes from FILE", "random-source", 0, "FILE");
+    args_parser.add_positional_argument(paths, "Path(s) to overwrite", "FILE", Core::ArgsParser::Required::Yes);
+    args_parser.parse(arguments);
+
+    auto rng_file = TRY(Core::File::open(random_source.value_or("/dev/random"sv), Core::File::OpenMode::Read));
+
+    for (auto& path : paths) {
+        auto file = TRY(Core::File::open(path, Core::File::OpenMode::ReadWrite | Core::File::OpenMode::DontCreate));
+
+        off_t output_file_length = TRY(FileSystem::size_from_fstat(file->fd()));
+
+        for (u32 iter = 0; iter < iterations; iter++) {
+            if (verbose)
+                outln("shred: {}: pass {}/{} (random)", path, iter + 1, iterations);
+
+            uint8_t buffer[BUFSIZ];
+            off_t total_written = 0;
+            ssize_t nread = 0;
+            ssize_t nwritten = 0;
+
+            while (total_written < output_file_length) {
+                size_t to_write_next = min(output_file_length - total_written, BUFSIZ);
+
+                nread = read(rng_file->fd(), &buffer, to_write_next);
+                if (nread < 0) {
+                    perror("read");
+                    return 1;
+                }
+
+                if (nread == 0)
+                    break;
+
+                nwritten = write(file->fd(), &buffer, to_write_next);
+                if (nwritten < 0) {
+                    perror("write");
+                    return 1;
+                }
+
+                if (nwritten == 0)
+                    break;
+
+                total_written += nwritten;
+            }
+
+            TRY(file->seek(0, SeekMode::SetPosition));
```

Comment:
Shouldn't we also perform a sync after each iteration so write caching (if any) doesn't prevent data from being written to the storage?